zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java

```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
         .extracting(Record::getValue)
         .extracting(ProcessInstanceRecordValue::getFlowScopeKey)
         .describedAs("Expect each B Task to be activated inside specific flow scopes")
-        .containsExactlyInAnyOrder(
-            aTasks.get(0).getValue().getFlowScopeKey(), aTasks.get(2).getValue().getFlowScopeKey());
+        .containsExactlyInAnyOrder(miInstances.getFirst().getKey(), miInstances.getLast().getKey());
+  }
+
+  @Test
+  public void shouldUseAncestorSelectionInsideMultiInstanceAdHocSubprocess() {
+    // given
+    ENGINE
+        .deployment()
+        .withXmlResource(
+            Bpmn.createExecutableProcess(PROCESS_ID)
+                .startEvent()
+                .adHocSubProcess(
+                    "adhoc_subprocess",
+                    adHocSubProcess -> {
+                      adHocSubProcess
+                          .multiInstance()
+                          .zeebeInputCollectionExpression("activateElements")
+                          .zeebeInputElement("element");
+                      adHocSubProcess.zeebeActiveElementsCollectionExpression("[element]");
+                      adHocSubProcess.userTask("A");
+                      adHocSubProcess.userTask("B");
+                    })
+                .endEvent()
+                .done())
+        .deploy();
+
+    final var processInstanceKey =
+        ENGINE
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariables(Map.of("activateElements", List.of("A", "A")))
+            .create();
+
+    // Get all multi-instance ad-hoc subprocess instances
+    final var miInnerInstances =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.AD_HOC_SUB_PROCESS_INNER_INSTANCE)
+            .limit(2)
+            .toList();
+    assertThat(miInnerInstances).hasSize(2);
+
+    final var aTasks =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementType(BpmnElementType.USER_TASK)
+            .withElementId("A")
+            .limit(2)
+            .toList();
+    assertThat(aTasks).hasSize(2);
+
+    // when
+    // terminate task A in the first ad-hoc instance and activate task B there
+    ENGINE
+        .processInstance()
+        .withInstanceKey(processInstanceKey)
+        .modification()
+        .terminateElement(aTasks.getFirst().getKey())
+        .activateElement("B", miInnerInstances.getFirst().getKey())
```

Comment:
🔧Here you are choosing the first ad hoc sub process' inner instance, assuming that it's the same as the flow scope of the task instance that you're terminating. Luckily that's the case, because the ordering is like that on the log. However, it would be better to be more explicit about the activation. What we want to do is move A to B within it's flowscope. Like you say in the comment "terminate task A in the first ad-hoc instance and activate task B there", so let's choose to activate B inside the same flowscope explicitly:
```suggestion
        .activateElement("B", aTasks.getFirst().getValue().getFlowScopeKey())
```